src/protections/DetectFederationLag.ts

```diff
@@ -0,0 +1,744 @@
+     * some stats otherwise.
+     */
+    public globalStats(): Stats | null {
+        return this.totalLag.stats();
+    }
+
+    /**
+     * Check if a server is currently marked as lagging.
+     *
+     * A server is marked as lagging if its mean lag has exceeded
+     * `threshold.enterWarningZone` and has not decreased below
+     * `threshold.exitWarningZone`.
+     *
+     * @returns `true` is that server is currently on alert.
+     */
+    public isServerOnAlert(serverId: string): boolean {
+        return this.serverAlerts.has(serverId);
+    }
+
+    /**
+     * The list of servers currently on alert.
+     */
+    public serversOnAlert(): IterableIterator<string> {
+        return this.serverAlerts.keys();
+    }
+
+    public cleanup(settings: HistogramSettings, now: Date, oldest: Date) {
+        // Cleanup global histogram.
+        //
+        // If `oldest == now - settings.duration * settings.number`, this
+        // should correspond exactly to the cleanup that takes place within
+        // `this.serverLags`. There is a risk of inconsistency between data
+        // if this is not the case.
+        //
+        // We assume that this is an acceptable risk: as we regularly
+        // erase oldest data from both `this.totalLag` and individual
+        // entries of `this.serverLags`, both sets of data will eventually
+        // catch up with each other.
+        this.totalLag.updateSettings(settings, now);
+        let serverLagsDeleteIds = [];
+        for (let [serverId, serverStats] of this.serverLags) {
+            if (serverStats.latestMessage < oldest) {
+                // Remove entire histogram.
+                serverLagsDeleteIds.push(serverId);
+                continue;
+            }
+            // Cleanup histogram.
+            serverStats.updateSettings(settings, now);
+        }
+        for (let key of serverLagsDeleteIds) {
+            this.serverLags.delete(key);
+            this.serverAlerts.delete(key);
+            // Note that we remove the alert to save memory (it's not really useful
+            // to keep monitoring a server for too long after receiving a message)
+            // but this does NOT guaranteed that server lag is over. It may be that
+            // the server is down or that the server is lagging by more than ~1h
+            // (by default).
+        }
+    }
+}
+
+export class DetectFederationLag extends Protection {
+    /**
+     * For each room we're monitoring, lag information.
+     */
+    lagPerRoom: Map<string /* roomId */, RoomInfo> = new Map();
+    public settings = {
+        // Rooms to ignore.
+        ignoreRooms: new StringSetProtectionSetting(),
+        // Servers to ignore, typically because they're known to be slow.
+        ignoreServers: new StringSetProtectionSetting(),
+        // How often we should recompute lag (ms).
+        bucketDurationMS: new NumberProtectionSetting(DEFAULT_BUCKET_DURATION_MS, 100),
```

Comment:
perhaps we should have a `DurationProtectionSetting` so people changing the option specify `100ms` rather than just `100`, that way the setting name doesn't have to explain what unit it is

Reply:
Done.